src/software/new_geom/linear_spline2d.cpp

```diff
@@ -0,0 +1,103 @@
+#include "software/new_geom/linear_spline2d.h"
+
+#include <algorithm>
+#include <sstream>
+
+LinearSpline2d::LinearSpline2d(const std::vector<Point>& points) : knots(points)
+{
+    initLinearSegments(points);
+}
+
+LinearSpline2d::LinearSpline2d(const std::initializer_list<Point>& points) : knots(points)
+{
+    initLinearSegments(points);
+}
+
+const Point LinearSpline2d::getValueAt(double val) const
+{
+    if (val < 0.0 || val > 1.0)
+    {
+        std::stringstream ss;
+        ss << "Tried to evaluate spline at " << val
+           << ", which is outside of domain of the spline: [0,1]";
+        throw std::invalid_argument(ss.str());
+    }
+
+    Point retval;
+
+    if (segments.empty())
+    {
+        retval = knots.front();
+    }
+    else
+    {
+        // Note: this could be more performant with binary search
+        auto seg_it = std::find_if(
+            segments.begin(), segments.end(), [val](const SplineSegment2d& sseg) {
+                return (val >= sseg.getStartVal() && val <= sseg.getEndVal());
+            });
+
+        if (seg_it == segments.end())
+        {
+            std::stringstream ss;
+            ss << "Tried to evaluate spline at " << val
+               << ", which was not in any interval of any segment";
+            throw std::runtime_error(ss.str());
+        }
+
+        retval = seg_it->getPolynomial().valueAt(val);
+    }
+
+    return retval;
+}
+
+size_t LinearSpline2d::getNumKnots(void) const
+{
+    return knots.size();
+}
+
+const std::vector<Point> LinearSpline2d::getKnots(void) const
+{
+    return knots;
+}
+
+const Point LinearSpline2d::getStartPoint(void) const
+{
+    return knots.front();
+}
+
+const Point LinearSpline2d::getEndPoint(void) const
+{
+    return knots.back();
+}
+
+const std::vector<SplineSegment2d> LinearSpline2d::getSplineSegments() const
+{
+    return segments;
+}
+
+void LinearSpline2d::initLinearSegments(const std::vector<Point>& points)
+{
+    if (points.size() == 0)
```

Comment:
Not handling the case when `points.size() == 1`

Reply:
What do we need to do in that case? the `points` are already set in the constructor and there are no segments since we only have a single point.

Reply:
Oh I guess that is a valid spline, just confused me when I read it